AMM/SPECIFICATION.md

```diff
@@ -0,0 +1,180 @@
+Table of Contents
+- [Overview](#overview)
+- [Use Cases](#use-cases)
+    - [Actions that users are able to perform](#actions-that-users-are-able-to-perform)
+        - [AMM Core Functionality](#amm-core-functionality)
+            - [`initialize()`](#initialize)
+            - [`add_pool()`](#add_pool)
+        - [AMM State Checks](#amm-state-checks)
+            - [`pool()`](#pool)
+        - [Exchange Core Functionality](#exchange-core-functionality)
+            - [`constructor()`](#constructor)
+            - [`deposit()`](#deposit)
+            - [`add_liquidity()`](#add_liquidity)
+            - [`remove_liquidity()`](#remove_liquidity)
+            - [`withdraw()`](#withdraw)
+            - [`swap_with_exact_input()`](#swap_with_exact_input)
+            - [`swap_with_exact_output()`](#swap_with_exact_output)
+        - [Exchange Previews](#exchange-previews)
+            - [`preview_add_liquidity()`](#preview_add_liquidity)
+            - [`preview_swap_with_exact_input()`](#preview_swap_with_exact_input)
+            - [`preview_swap_with_exact_output()`](#preview_swap_with_exact_output)
+        - [Exchange State Checks](#exchange-state-checks)
+            - [`balance()`](#balance)
+            - [`pool_info()`](#pool_info)
+- [Sequence Diagram](#sequence-diagram)
+
+# Overview
+
+This document provides an overview of the application.
+
+It outlines the use cases, i.e. desirable functionality, in addition to requirements for the smart contracts.
+
+# Use Cases
+
+This section contains general information about the functionality of the application and thus does not touch upon any technical aspects.
+
+If you are interested in a functional overview then this is the section for you.
+
+## Actions that users are able to perform
+
+This sub-section details what a user is able to do e.g. click a button and "x, y, z" happens.
+
+### AMM Core Functionality
+
+#### `initialize()`
+
+1. Allows specifying the legitimate exchange contract implementation that the AMM will operate with Requires
+    1. The bytecode root of the desired exchange contract implementation
+
+#### `add_pool()`
+
+1. Allows an asset pair to be registered to the AMM 
+    1. If the AMM is initialized Requires
+    2. Requires
+        1. The identifiers of the two assets
+        2. The exchange contract identifier that is also the identifier of the liquidity pool asset for the given pair if the exchange contract
+            1. Is legitimate, i.e., the bytecode root matches
+            2. Defines the pool for the provided asset pair
```

Comment:
These should be reworded to be separate points rather than a `requires` "tag" somewhere which the user needs to look for and figure out some sentence substitution. Each bullet point should be self contained and explicit